test/integration/integration_test.go

```diff
@@ -0,0 +1,203 @@
+	secret, err := msal.CreateClientCredentialFromSecret(secretS)
+	if err != nil {
+		return nil, err
+	}
+	options := msal.DefaultConfidentialClientApplicationOptions()
+	clientID := os.Getenv("clientId")
+	options.Authority = microsoftAuthority
+	app, err := msal.NewConfidentialClientApplication(clientID, secret, &options)
+	if err != nil {
+		return nil, err
+	}
+	return &labClient{app: app}, nil
+}
+func (l *labClient) getLabAccessToken() (string, error) {
+	scopes := []string{msIDlabDefaultScope}
+	result, err := l.app.AcquireTokenSilent(context.Background(), scopes, nil)
+	if err != nil {
+		result, err = l.app.AcquireTokenByClientCredential(context.Background(), scopes)
+		if err != nil {
+			return "", err
+		}
+	}
+	return result.GetAccessToken(), nil
+}
+
+func (l *labClient) getUser(query map[string]string) (user, error) {
+	accessToken, err := l.getLabAccessToken()
+	if err != nil {
+		return user{}, err
+	}
+	q := url.Values{}
+	for key, value := range query {
+		q.Add(key, value)
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/user", q, accessToken)
+	if err != nil {
+		return user{}, err
+	}
+	var users []user
+	err = json.Unmarshal(responseBody, &users)
+	if err != nil {
+		return user{}, err
+	}
+	if len(users) == 0 {
+		return user{}, errors.New("No user found")
+	}
+	user := users[0]
+	user.Password, err = l.getSecret(map[string]string{"Secret": user.LabName})
+	if err != nil {
+		return user, err
+	}
+	return user, nil
+}
+
+func (l *labClient) getSecret(query map[string]string) (string, error) {
+	accessToken, err := l.getLabAccessToken()
+	if err != nil {
+		return "", err
+	}
+	q := url.Values{}
+	for key, value := range query {
+		q.Add(key, value)
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/LabUserSecret", q, accessToken)
+	if err != nil {
+		return "", err
+	}
+	var secret secret
+	err = json.Unmarshal(responseBody, &secret)
+	if err != nil {
+		return "", err
+	}
+	return secret.Value, nil
+}
+
+// TODO: Add getApp() when needed
+
+func getTestUser(lc *labClient, query map[string]string) user {
+	testUser, err := lc.getUser(query)
+	if err != nil {
+		panic("Failed to get input user. " + err.Error())
+	}
+	return testUser
+}
+
+func TestUsernamePassword(t *testing.T) {
+	labClientInstance, err := newLabClient()
+	if err != nil {
+		panic("Failed to get a lab client. " + err.Error())
+	}
+	tests := []struct {
+		desc string
+		user user
+	}{
+		{"Managed", getTestUser(labClientInstance, map[string]string{"usertype": "cloud"})},
+		{"ADFSv2", getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv2"})},
+		{"ADFSv3", getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv3"})},
+		{"ADFSv4", getTestUser(labClientInstance, map[string]string{"usertype": "federated", "federationProvider": "ADFSv4"})},
+	}
+	for _, test := range tests {
+		options := msal.DefaultPublicClientApplicationOptions()
```

Comment:
Sorry, just now noticed this:

From line 181 to line 186: You can move this outside the test loop.  This is the same client and option with no internal state changes, so we only need to create it once.

Reply:
Not all of 181-186 could be put outside because the appId used to initialize the public client depends on the testcase, but I have moved the common part out